### ISSUE TITLE
amp-lightbox-gallery: modal dialog issues

### ISSUE THREAD

**Issue by TetraLogicalHelpdesk** [ampproject/amphtml#31183]
## WCAG Level
Level A

## Priority
High

## Type
Navigation, Semantics

## Pages/screens/components affected
* [amp-lightbox-gallery](https://amp.dev/documentation/components/amp-lightbox-gallery/)

## Description
If the focus order is not logical, sighted keyboard users are likely to be confused and irritated when navigating through a site, particularly if the focus appears to "jump" in an illogical manner to different, seemingly unconnected parts of the page. This issue will be even more problematic of users of assistive technologies, who may not have the benefit of being able to perceive the overall layout/structure of the page to realize that that the illogical focus order has potentially moved them to a completely separate part of the current page.

When focus is not programmatically moved to the most appropriate location following an interaction, users will be left to manually navigate again to the most appropriate location in the page. For sighted keyboard users, this can be an inconvenience/irritation, particularly if focus was lost/reset to the start of the document and they have to cycle through the entire page again. Blind and low vision assistive technology users may be unaware of what change actually occurred on the page as a result of their interaction. After performing an action, these users may be confused/left wondering if their action was actually successful. In cases where focus is lost (generally caused by the element that previously had focus being disabled or removed from the page), some browsers may error-correct and attempt to relocate the user's focus to the nearest focusable element, or they may simply reset the focus to the start of the document. Overall, it will likely make the user's experience of navigating the page unnecessarily tedious and confusing.

While, visually, `amp-lightbox-gallery` is presented as a modal dialog, it currently does not have any of the necessary focus behaviors of a modal dialog. Note that currently, the `amp-lightbox-gallery` is also completely inaccessible to keyboard users (see separate issue).

When the dialog is opened, focus is not moved explicitly to the dialog. Focus is not retained inside the modal dialog - elements in the underlying page, behind the semi-transparent overlay of the dialog, are still focusable, as well as being exposed to assistive technologies. When the dialog is closed, focus is lost/reset.

In addition, the modal dialog itself is not programmatically identified as a modal dialog to assistive technologies.

## User impact
When the dialog is displayed, keyboard users will have to explicitly navigate through the (now visually hidden/dimmed) underlying page in order to reach the dialog. Once inside the dialog, users can end up navigating back out of the dialog and back into the underlying page. Once the dialog is closed, keyboard focus is lost (since the last focused element inside the dialog is removed). While most browsers error-correct for this, in certain situations this can result in keyboard focus returning to the start of the current page.

For assistive technology users, it will not be apparent that a modal dialog was opened. The dialog itself lacks any programmatic indication that it is in fact a modal dialog, and the underlying page - though visually hidden behind the semi-transparent overlay - is still accessible, giving no indication that it is currently inoperable.

## Required solution
Make sure that the order in which elements receive keyboard focus follows a meaningful and logical sequence. The order does not necessarily have to follow the order implied by the visual layout, as long as it is logical. Nonetheless, the focus should ideally not "jump" erratically between different parts of the layout.

Pay attention to source order. Sequencing elements in the code so they follow a logical order helps to provide a logical and meaningful focus order.

In the case of modal dialogs, make sure that - once the modal dialog is displayed - focus is programmatically moved to the dialog. Make sure that focus remains inside the modal dialog. Once the dialog is closed, explicitly move focus back to the most appropriate element in the page - generally, the element that last had focus when the dialog was opened / the element that triggered the modal dialog in the first place.

To convey the dialog appropriately to assistive technologies, add `role="dialog"` and `aria-modal="true"` to the dialog container itself. For backwards compatibility with older assistive technologies, we further recommend explicitly hiding the underlying page using `aria-hidden="true"`.

Refer to the [WAI-ARIA Authoring Practices 1.1 dialog section](https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal) for further details.

This solution must be applied to all instances of the issue identified within the test sample, then applied to all other instances of the same issue identified throughout the rest of the website.

### Initial focus for complex dialogs

For complex dialogs - particularly those that contain a mixture of descriptive text and additional form controls, or structured information/content - we recommend moving focus programmatically to the start of the modal dialog itself (for instance, the actual dialog container).

Add a `tabindex="-1"` to the container of the modal, to make it programmatically focusable, and then set `focus()` to it when it is displayed. For these types of dialogs, omit the `aria-labelledby` and `aria-describedby` attributes mentioned in the previous section, as otherwise assistive technology users will hear the content twice (once when focus lands on the modal dialog itself, and then when they actually navigate through the dialog).

### Hiding content in the underlying page from assistive technologies and preventing keyboard focus

When a modal dialog is displayed, any content that is not part of the modal dialog content must not be included in the focus order (since the dialog should "contain" focus). However, content should also be hidden from assistive technologies, to avoid that AT users navigate back into the underlying main document using alternative methods (such as the virtual cursor, navigation by headings, etc.).

As a general approach, use `aria-hidden="true"` to hide the main content. Structurally, this will require a change to the page structure, as the modal dialog itself then cannot be a child element of the regular container (which will be hidden from assistive technologies). Conceptually, this means that modal dialogs should live "outside" of the element that holds the main content of the page.

```
<body>
  <div aria-hidden="true">
    <!-- page content not in modal dialog -->
  </div>
  <div role="dialog">
    <!-- dialog content -->
  </div>
</body>
```

Make sure to remove `aria-hidden="true"` once the modal dialog is closed.

Hiding content from assistive technologies this way still won't prevent any focusable elements from receiving focus. To achieve this, you may potentially need to programmatically add `tabindex="-1"` to each focusable element in the content (and then remove these again when the modal dialog is closed).

For a solid implementation of an accessible custom dialog (which hides content from the underlying page and makes it unfocusable), see [A11y Dialog](https://github.com/edenspiekermann/a11y-dialog).

### Focus handling within the dialog

The `amp-lightbox-gallery` can be switched from between a "Gallery" and "Content" view. Make sure that when a user switches between these modes, focus is correctly moved explicitly to the newly shown modal content, and that the now hidden controls are not keyboard-focusable or exposed to assistive technologies.

## Test procedure(s)
Use these steps to confirm that the solution has been correctly applied to issues identified within the test sample, and to test the rest of the website for instances of the same issue:
1. Open a modal dialog.
2. Verify that focus has been moved to the modal dialog (and is not left in the underlying page).
3. Verify that focus remains inside the modal dialog - using `Tab`/`Shift`+`Tab`, focus should cycle through the focusable elements of the modal itself and not return to the underlying page.
4. Move focus to the browser's address bar (`F6` or `Alt`+`D`), then proceed forward with `Tab`; verify that focus only goes to the elements in the modal dialog, and not to elements in the underlying page.
5. Using assistive technology, or developer tools, verify that the underlying page has been hidden from assistive technologies (uses `aria-hidden="true"` on the page content container).
6. Close the modal dialog.
7. Verify that focus has returned correctly to the most appropriate place in the page - generally, the control that triggered the modal in the first place.
8. Using assistive technology, or developer tools, verify that the page is not hidden from assistive technologies anymore (no presence of `aria-hidden="true"` on the page content container).

### Definition of done
Complete all of these tasks before closing this issue or indicating it is ready for retest:
* All issues identified within the test sample have been resolved.
* The rest of the website has been tested for the same issue.
* All issues identified throughout the rest of the website have been resolved or filed as new issues.

## Related standards
* [WCAG 2.1 Success Criterion 1.3.1 Info and Relationships (Level A)](https://www.w3.org/TR/WCAG21/#info-and-relationships)
* [WCAG 2.1 Success Criterion 1.3.2 Meaningful Sequence (Level A)](https://www.w3.org/TR/WCAG21/#meaningful-sequence)
* [WCAG 2.1 Success Criterion 2.4.3 Focus Order (Level A)](https://www.w3.org/TR/WCAG21/#focus-order)
* [WCAG 2.1 Success Criterion 4.1.2 Name, Role, Value (Level A)](https://www.w3.org/TR/WCAG21/#name-role-value)

## More information
* [W3C - Placing the interactive elements in an order that follows sequences and relationships within the content](https://www.w3.org/WAI/WCAG21/Techniques/general/G59)
* [WAI-ARIA Authoring Practices 1.1 dialog section](https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal)
* [A11y Dialog](https://github.com/edenspiekermann/a11y-dialog)

## Test data
Test date: November 2020
